### PR TITLE
feat(aws-sns-sqs) updated sqs helper function to support fifo type queues

### PR DESCRIPTION
**Pull request by mickychetta**
*Issue #342 , if available:*

*Description of changes:*
Error was coming from the DLQ not being the same type as the source queue
-Updated the sqs-helper function to have fifo prop, if it is undefined then it would be a standard queue
-fifo prop is of type boolean but if it is `false` it raises an UnknownAttribute error therefore it is either `true` or `undefined`: https://github.com/aws/aws-cdk/issues/8550

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: d3528072a5e67414ad3f7939ccf0f8b3c4073b41
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bf0324ed-694c-4016-9b6c-77b62b1fbb87%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/core/lib/sqs-helper.ts [hunk 1]
```diff
@@ -125,13 +125,18 @@ export interface BuildDeadLetterQueueProps {
...
+   */
+  readonly fifo?: boolean
 }
 
 export function buildDeadLetterQueue(scope: cdk.Construct, props: BuildDeadLetterQueueProps): sqs.DeadLetterQueue | undefined {
   if (!props.existingQueueObj && (props.deployDeadLetterQueue || props.deployDeadLetterQueue === undefined)) {
     // Create the Dead Letter Queue
     const [dlq] = buildQueue(scope, 'deadLetterQueue', {
-      queueProps: props.deadLetterQueueProps
+      queueProps: { ...props.deadLetterQueueProps, fifo: props.fifo ? true : undefined }
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/core/lib/sqs-helper.ts:139]: Do you need this line? Line 132 already has already made fifo an attribute of deatLetterQueueProps.

**Review by biffgaut — CHANGES_REQUESTED** [review]
We need at least 1 unit test and 1 integration that covers this.
Did you actually run this or are we thinking this works correctly by inspection? I ask because of the lack of a test - I would have started by writing a test that broke, then making my changes so the test worked.

File: source/patterns/@aws-solutions-constructs/core/lib/sqs-helper.ts (hunk 1 above)
Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/core/lib/sqs-helper.ts:139]: Yes, deadLetterQueueProps has the fifo property but this ensures it is passed properly depending on the logic:

Approach 1:
```
// Pass prop like this
this.deadLetterQueue = defaults.buildDeadLetterQueue(this, {
  existingQueueObj: props.existingQueueObj,
  deployDeadLetterQueue: props.deployDeadLetterQueue,
  deadLetterQueueProps: props.deadLetterQueueProps,
  maxReceiveCount: props.maxReceiveCount,
  fifo: props.queueProps && props.queueProps.fifo
});

// Build DLQ like this 
const [dlq] = buildQueue(scope, 'deadLetterQueue', {
      queueProps: { ...props.deadLetterQueueProps, fifo: props.fifo }
    });
```

OR 

Approach 2:
```
// Pass prop like this
this.deadLetterQueue = defaults.buildDeadLetterQueue(this, {
  existingQueueObj: props.existingQueueObj,
  deployDeadLetterQueue: { ...props.deployDeadLetterQueue, fifo: props.queueProps && props.queueProps.fifo },
  deadLetterQueueProps: props.deadLetterQueueProps,
  maxReceiveCount: props.maxReceiveCount
});

// Build DLQ like this 
const [dlq] = buildQueue(scope, 'deadLetterQueue', {
  queueProps: { ...props.deadLetterQueueProps }
});
```
I think the first approach is cleaner, what did you think?

Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/core/lib/sqs-helper.ts]: Manually tested before but now updated with integ and unit tests

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: e47d7103557c040bde6be4f2e848c803c0bef36b
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8b4e3dfe-c3be-46e4-aead-640ea038be75%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: 16f0e7c57e9195ad086509e4b5191a5202e61caf
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a051a1b0-eed5-44b3-a46e-b39c8facc4d1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: 596e6354821be56f5731a8c87763f52fbf738f5f
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=45073c96-2ab0-47a0-89bf-e58709bc544d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: 08f1f918c4fc6030e2e50ca15e740977728ba95d
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d560f273-c307-48e6-93a7-0be8baa7cbb1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: a7c709b37a008c262964b3943b4c5beac025d41f
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9ca4e1e9-e457-40bf-a31f-d998d77905d6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts [hunk 2]
```diff
@@ -92,6 +92,23 @@ export function CheckProps(propsObject: VerifiedProps | any) {
     errorFound = true;
   }
 
+  if (propsObject.queueProps && propsObject.queueProps.fifo && propsObject.deployDeadLetterQueue === undefined &&
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts:95]: propsObject.queueProps && propsObject.queueProps.fifo can be shortened to

propsObject?.queueProps?.fifo

Lines 95, 96, 101 and 102 are extremely intimidating and confusing overall, I'm going to look for a simplification...

Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts:95]: How about replacing 95-105 with:
```
const isQueueFifo: boolean = propsObject?.queueProps?.fifo;
const isDeadLetterQueueFifo: boolean = propsObject?. deadLetterQueueProps?.fifo;

if (isQueueFifo != isDeadLetterQueueFifo) {
    errorMessages += 'Error - If you specify a fifo: true in either queueProps or deadLetterQueueProps, you must also set fifo: true in the other props object. Fifo must match for the Queue and the Dead Letter Queue.\n';
    errorFound = true;
}
```
Check me out to ensure this logic is correct. The ? operator is most unfamiliar to me, that's where I may have screwed up - if they are specified correctly the actual if statement is very clear.

**Review by biffgaut — CHANGES_REQUESTED** [review]
Nice work adding some fifo tests. I didn't read them carefully, do we have one that specifically triggers the new input validation?

Lines like 105-115 in input-validation.ts should always raise red flags - 

* The if statements are very complicated
* Confirming all the possible combinations is difficult (and therefore error prone)
* The same decisions appear multiple times across the two if statements

There's no rule, nor even benefit to fitting everything into a giant if statement. Breaking it down and pulling out smaller hunks to assign a meaningful name can make it much easier to read. I used a Typescript feature and a couple variables in a suggested replacement that I think you agree is much more readable (and easier to verify). That said, please examine it closely and confirm that my logic is correct.

File: source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts [hunk 3]
```diff
@@ -92,6 +92,23 @@ export function CheckProps(propsObject: VerifiedProps | any) {
...
+    errorFound = true;
+  }
+
+  if (propsObject.deadLetterQueueProps && propsObject.deadLetterQueueProps.fifo && propsObject.deployDeadLetterQueue === undefined &&
+    ((!propsObject.queueProps) || (propsObject.queueProps && !propsObject.queueProps.fifo))) {
+    errorMessages += 'Error - If you specify a fifo: true in deadLetterQueueProps, you must also supply queueProps with fifo: true if deploying a Dead Letter Queue.\n';
+    errorFound = true;
+  }
+
+  if (propsObject.existingQueueObj && propsObject.queueProps) {
```
[the remark this answers is not in the document]
Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts:107]: Yep we have some that trigger the new input validation. Also the lines were duplicates, I just deleted them.

File: source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts (hunk 2 above)
Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts:95]: This works good, much more cleaner

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: 293cff5e897288c33d59e44416569752539f99da
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d8ab1a28-ba7e-403c-9214-5c81435166bd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts [hunk 4]
```diff
@@ -71,6 +71,9 @@ export interface VerifiedProps {
 export function CheckProps(propsObject: VerifiedProps | any) {
   let errorMessages = '';
   let errorFound = false;
+  const isQueueFifo: boolean = propsObject?.queueProps?.fifo;
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts:74]: There's no rule that these must be near the top of the routine. Their only purpose is to introduce names that add clarity to the decisions in line 108 - go ahead and put them just above that decision.

**Review by biffgaut — CHANGES_REQUESTED** [review]

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: 36f05f772e4b034a798481a1d1e26a6278383f31
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=532d02d6-5a72-4bca-a17e-867431403f0d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by biffgaut — APPROVED** [review]

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: 48f2069a6acb5388494e95977310779c6e41e899
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=532d6816-a09d-41af-8008-c959ac1995f5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by biffgaut — APPROVED** [review]

**Pull request merged**